{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          Transcript ID  Position Nucleotide 1 Nucleotide 2 Nucleotide 3  \\\n",
      "0       ENST00000000233       244        AAGAC        AGACC        GACCA   \n",
      "1       ENST00000000233       261        CAAAC        AAACT        AACTG   \n",
      "2       ENST00000000233       316        GAAAC        AAACA        AACAG   \n",
      "3       ENST00000000233       332        AGAAC        GAACA        AACAT   \n",
      "4       ENST00000000233       368        AGGAC        GGACA        GACAA   \n",
      "...                 ...       ...          ...          ...          ...   \n",
      "121833  ENST00000641834      1348        GGGAC        GGACA        GACAT   \n",
      "121834  ENST00000641834      1429        CTGAC        TGACA        GACAC   \n",
      "121835  ENST00000641834      1531        TGGAC        GGACA        GACAC   \n",
      "121836  ENST00000641834      1537        CTGAC        TGACC        GACCA   \n",
      "121837  ENST00000641834      1693        TTGAC        TGACA        GACAT   \n",
      "\n",
      "        N1 Length Avg  N1 Length Median  N1 Length Min  N1 Length Max  \\\n",
      "0            0.008264           0.00697        0.00199         0.0339   \n",
      "1            0.006609           0.00564        0.00199         0.0222   \n",
      "2            0.007570           0.00631        0.00232         0.0299   \n",
      "3            0.010620           0.00902        0.00232         0.0370   \n",
      "4            0.010701           0.00896        0.00199         0.0478   \n",
      "...               ...               ...            ...            ...   \n",
      "121833       0.009594           0.00817        0.00232         0.0432   \n",
      "121834       0.008393           0.00618        0.00232         0.0262   \n",
      "121835       0.008161           0.00697        0.00232         0.0315   \n",
      "121836       0.008044           0.00660        0.00232         0.0483   \n",
      "121837       0.008788           0.00709        0.00232         0.0215   \n",
      "\n",
      "        N1 SD Avg  ...  N3 Length Min  N3 Length Max  N3 SD Avg  N3 SD Median  \\\n",
      "0        4.223784  ...        0.00232         0.0329   4.386989         3.440   \n",
      "1        3.216424  ...        0.00232         0.0262   3.016599         2.660   \n",
      "2        2.940541  ...        0.00199         0.0266   2.087146         1.910   \n",
      "3        6.476350  ...        0.00199         0.0214   2.236520         2.135   \n",
      "4        6.415051  ...        0.00266         0.0485   4.260253         4.160   \n",
      "...           ...  ...            ...            ...        ...           ...   \n",
      "121833   3.294164  ...        0.00232         0.0242   4.005616         3.640   \n",
      "121834   4.511014  ...        0.00232         0.0149   3.644638         3.440   \n",
      "121835   3.918438  ...        0.00232         0.0260   2.181562         1.925   \n",
      "121836   3.191228  ...        0.00258         0.0144   2.540877         2.330   \n",
      "121837   4.090577  ...        0.00232         0.0306   2.576731         2.295   \n",
      "\n",
      "        N3 SD Min  N3 SD Max  N3 Mean Avg  N3 Mean Median  N3 Mean Min  \\\n",
      "0           0.773      15.50    80.570270           80.50         73.1   \n",
      "1           0.715      14.10    94.290698           94.10         88.6   \n",
      "2           0.630       6.85    89.364324           89.20         84.4   \n",
      "3           0.884       6.49    89.154000           89.90         81.4   \n",
      "4           1.040       8.81    85.178788           85.40         77.6   \n",
      "...           ...        ...          ...             ...          ...   \n",
      "121833      1.030      12.50    82.004110           82.10         74.0   \n",
      "121834      1.580       6.50    80.497101           80.60         75.0   \n",
      "121835      1.040       7.19    84.190625           84.50         78.2   \n",
      "121836      1.190       6.50    82.289474           82.00         77.6   \n",
      "121837      1.330       5.67    78.536538           77.45         74.4   \n",
      "\n",
      "        N3 Mean Max  \n",
      "0              88.3  \n",
      "1             103.0  \n",
      "2              96.2  \n",
      "3              95.7  \n",
      "4              90.5  \n",
      "...             ...  \n",
      "121833         92.8  \n",
      "121834         90.6  \n",
      "121835         88.1  \n",
      "121836         87.4  \n",
      "121837         87.6  \n",
      "\n",
      "[121838 rows x 41 columns]\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "import os\n",
    "from io import StringIO\n",
    "import gzip\n",
    "import json\n",
    "import numpy as np\n",
    "from decimal import Decimal, getcontext\n",
    "# Prepare a list to collect rows\n",
    "\n",
    "cwd = os.getcwd() # Gets current working directory; ensure that .json file is in the same folder as this script.\n",
    "\n",
    "path = os.path.join(cwd, 'dataset0.json.gz')\n",
    "labels_path = os.path.join(cwd, 'data.info.labelled') # replace with your actual file path\n",
    "\n",
    "aggregated_data = []\n",
    "\n",
    "# Read the gzipped JSON file line by line\n",
    "with gzip.open(file_path, 'rt') as f:\n",
    "    for line in f:\n",
    "        # Each line is a JSON object\n",
    "        data = json.loads(line)\n",
    "        for transcript_id, positions in data.items():\n",
    "            for pos, flanking_data in positions.items():\n",
    "                for sequence, reads in flanking_data.items():\n",
    "                    # Convert the list of reads to a NumPy array for aggregation\n",
    "                    features_array = np.array(reads)\n",
    "                    \n",
    "                    # Split the sequence into three nucleotides\n",
    "                    nucleotide1 = sequence[0:5]\n",
    "                    nucleotide2 = sequence[1:6]\n",
    "                    nucleotide3 = sequence[2:7]\n",
    "\n",
    "                    # Calculate mean, median, min, and max for each feature\n",
    "                    aggregated = {\n",
    "                        'mean': np.mean(features_array, axis=0).tolist(),\n",
    "                        'median': np.median(features_array, axis=0).tolist(),\n",
    "                        'min': np.min(features_array, axis=0).tolist(),\n",
    "                        'max': np.max(features_array, axis=0).tolist()\n",
    "                    }\n",
    "\n",
    "                    # Flatten the results into a single row\n",
    "                    row = {\n",
    "                        'Transcript ID': transcript_id,\n",
    "                        'Position': int(pos),\n",
    "                        'Nucleotide 1': nucleotide1,\n",
    "                        'Nucleotide 2': nucleotide2,\n",
    "                        'Nucleotide 3': nucleotide3\n",
    "                    }\n",
    "\n",
    "                    # Define the feature names\n",
    "                    feature_names = [\n",
    "                        'N1 Length', 'N1 SD', 'N1 Mean',\n",
    "                        'N2 Length', 'N2 SD', 'N2 Mean',\n",
    "                        'N3 Length', 'N3 SD', 'N3 Mean'\n",
    "                    ]\n",
    "                    \n",
    "                    for i, feature_name in enumerate(feature_names):\n",
    "                        row[f'{feature_name} Avg'] = aggregated['mean'][i]\n",
    "                        row[f'{feature_name} Median'] = aggregated['median'][i]\n",
    "                        row[f'{feature_name} Min'] = aggregated['min'][i]\n",
    "                        row[f'{feature_name} Max'] = aggregated['max'][i]\n",
    "\n",
    "                    aggregated_data.append(row)\n",
    "\n",
    "# Convert rows to a DataFrame\n",
    "\n",
    "data_df = pd.DataFrame(aggregated_data)\n",
    "\n",
    "# Display the DataFrame\n",
    "print(data_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>Position</th>\n",
       "      <th>Nucleotide 1</th>\n",
       "      <th>Nucleotide 2</th>\n",
       "      <th>Nucleotide 3</th>\n",
       "      <th>N1 Length Avg</th>\n",
       "      <th>N1 Length Median</th>\n",
       "      <th>N1 Length Min</th>\n",
       "      <th>N1 Length Max</th>\n",
       "      <th>...</th>\n",
       "      <th>N3 Length Max</th>\n",
       "      <th>N3 SD Avg</th>\n",
       "      <th>N3 SD Median</th>\n",
       "      <th>N3 SD Min</th>\n",
       "      <th>N3 SD Max</th>\n",
       "      <th>N3 Mean Avg</th>\n",
       "      <th>N3 Mean Median</th>\n",
       "      <th>N3 Mean Min</th>\n",
       "      <th>N3 Mean Max</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>244</td>\n",
       "      <td>AAGAC</td>\n",
       "      <td>AGACC</td>\n",
       "      <td>GACCA</td>\n",
       "      <td>0.008264</td>\n",
       "      <td>0.00697</td>\n",
       "      <td>0.00199</td>\n",
       "      <td>0.0339</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0329</td>\n",
       "      <td>4.386989</td>\n",
       "      <td>3.440</td>\n",
       "      <td>0.773</td>\n",
       "      <td>15.50</td>\n",
       "      <td>80.570270</td>\n",
       "      <td>80.50</td>\n",
       "      <td>73.1</td>\n",
       "      <td>88.3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>261</td>\n",
       "      <td>CAAAC</td>\n",
       "      <td>AAACT</td>\n",
       "      <td>AACTG</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>0.00564</td>\n",
       "      <td>0.00199</td>\n",
       "      <td>0.0222</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0262</td>\n",
       "      <td>3.016599</td>\n",
       "      <td>2.660</td>\n",
       "      <td>0.715</td>\n",
       "      <td>14.10</td>\n",
       "      <td>94.290698</td>\n",
       "      <td>94.10</td>\n",
       "      <td>88.6</td>\n",
       "      <td>103.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>316</td>\n",
       "      <td>GAAAC</td>\n",
       "      <td>AAACA</td>\n",
       "      <td>AACAG</td>\n",
       "      <td>0.007570</td>\n",
       "      <td>0.00631</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>0.0299</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0266</td>\n",
       "      <td>2.087146</td>\n",
       "      <td>1.910</td>\n",
       "      <td>0.630</td>\n",
       "      <td>6.85</td>\n",
       "      <td>89.364324</td>\n",
       "      <td>89.20</td>\n",
       "      <td>84.4</td>\n",
       "      <td>96.2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>332</td>\n",
       "      <td>AGAAC</td>\n",
       "      <td>GAACA</td>\n",
       "      <td>AACAT</td>\n",
       "      <td>0.010620</td>\n",
       "      <td>0.00902</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>0.0370</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0214</td>\n",
       "      <td>2.236520</td>\n",
       "      <td>2.135</td>\n",
       "      <td>0.884</td>\n",
       "      <td>6.49</td>\n",
       "      <td>89.154000</td>\n",
       "      <td>89.90</td>\n",
       "      <td>81.4</td>\n",
       "      <td>95.7</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000004059</td>\n",
       "      <td>ENST00000000233</td>\n",
       "      <td>368</td>\n",
       "      <td>AGGAC</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>GACAA</td>\n",
       "      <td>0.010701</td>\n",
       "      <td>0.00896</td>\n",
       "      <td>0.00199</td>\n",
       "      <td>0.0478</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0485</td>\n",
       "      <td>4.260253</td>\n",
       "      <td>4.160</td>\n",
       "      <td>1.040</td>\n",
       "      <td>8.81</td>\n",
       "      <td>85.178788</td>\n",
       "      <td>85.40</td>\n",
       "      <td>77.6</td>\n",
       "      <td>90.5</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121833</th>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1348</td>\n",
       "      <td>GGGAC</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>GACAT</td>\n",
       "      <td>0.009594</td>\n",
       "      <td>0.00817</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>0.0432</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0242</td>\n",
       "      <td>4.005616</td>\n",
       "      <td>3.640</td>\n",
       "      <td>1.030</td>\n",
       "      <td>12.50</td>\n",
       "      <td>82.004110</td>\n",
       "      <td>82.10</td>\n",
       "      <td>74.0</td>\n",
       "      <td>92.8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121834</th>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1429</td>\n",
       "      <td>CTGAC</td>\n",
       "      <td>TGACA</td>\n",
       "      <td>GACAC</td>\n",
       "      <td>0.008393</td>\n",
       "      <td>0.00618</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>0.0262</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0149</td>\n",
       "      <td>3.644638</td>\n",
       "      <td>3.440</td>\n",
       "      <td>1.580</td>\n",
       "      <td>6.50</td>\n",
       "      <td>80.497101</td>\n",
       "      <td>80.60</td>\n",
       "      <td>75.0</td>\n",
       "      <td>90.6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121835</th>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1531</td>\n",
       "      <td>TGGAC</td>\n",
       "      <td>GGACA</td>\n",
       "      <td>GACAC</td>\n",
       "      <td>0.008161</td>\n",
       "      <td>0.00697</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>0.0315</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0260</td>\n",
       "      <td>2.181562</td>\n",
       "      <td>1.925</td>\n",
       "      <td>1.040</td>\n",
       "      <td>7.19</td>\n",
       "      <td>84.190625</td>\n",
       "      <td>84.50</td>\n",
       "      <td>78.2</td>\n",
       "      <td>88.1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121836</th>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1537</td>\n",
       "      <td>CTGAC</td>\n",
       "      <td>TGACC</td>\n",
       "      <td>GACCA</td>\n",
       "      <td>0.008044</td>\n",
       "      <td>0.00660</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>0.0483</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0144</td>\n",
       "      <td>2.540877</td>\n",
       "      <td>2.330</td>\n",
       "      <td>1.190</td>\n",
       "      <td>6.50</td>\n",
       "      <td>82.289474</td>\n",
       "      <td>82.00</td>\n",
       "      <td>77.6</td>\n",
       "      <td>87.4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121837</th>\n",
       "      <td>ENSG00000167747</td>\n",
       "      <td>ENST00000641834</td>\n",
       "      <td>1693</td>\n",
       "      <td>TTGAC</td>\n",
       "      <td>TGACA</td>\n",
       "      <td>GACAT</td>\n",
       "      <td>0.008788</td>\n",
       "      <td>0.00709</td>\n",
       "      <td>0.00232</td>\n",
       "      <td>0.0215</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0306</td>\n",
       "      <td>2.576731</td>\n",
       "      <td>2.295</td>\n",
       "      <td>1.330</td>\n",
       "      <td>5.67</td>\n",
       "      <td>78.536538</td>\n",
       "      <td>77.45</td>\n",
       "      <td>74.4</td>\n",
       "      <td>87.6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>121838 rows Ã— 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Gene ID    Transcript ID  Position Nucleotide 1 Nucleotide 2  \\\n",
       "0       ENSG00000004059  ENST00000000233       244        AAGAC        AGACC   \n",
       "1       ENSG00000004059  ENST00000000233       261        CAAAC        AAACT   \n",
       "2       ENSG00000004059  ENST00000000233       316        GAAAC        AAACA   \n",
       "3       ENSG00000004059  ENST00000000233       332        AGAAC        GAACA   \n",
       "4       ENSG00000004059  ENST00000000233       368        AGGAC        GGACA   \n",
       "...                 ...              ...       ...          ...          ...   \n",
       "121833  ENSG00000167747  ENST00000641834      1348        GGGAC        GGACA   \n",
       "121834  ENSG00000167747  ENST00000641834      1429        CTGAC        TGACA   \n",
       "121835  ENSG00000167747  ENST00000641834      1531        TGGAC        GGACA   \n",
       "121836  ENSG00000167747  ENST00000641834      1537        CTGAC        TGACC   \n",
       "121837  ENSG00000167747  ENST00000641834      1693        TTGAC        TGACA   \n",
       "\n",
       "       Nucleotide 3  N1 Length Avg  N1 Length Median  N1 Length Min  \\\n",
       "0             GACCA       0.008264           0.00697        0.00199   \n",
       "1             AACTG       0.006609           0.00564        0.00199   \n",
       "2             AACAG       0.007570           0.00631        0.00232   \n",
       "3             AACAT       0.010620           0.00902        0.00232   \n",
       "4             GACAA       0.010701           0.00896        0.00199   \n",
       "...             ...            ...               ...            ...   \n",
       "121833        GACAT       0.009594           0.00817        0.00232   \n",
       "121834        GACAC       0.008393           0.00618        0.00232   \n",
       "121835        GACAC       0.008161           0.00697        0.00232   \n",
       "121836        GACCA       0.008044           0.00660        0.00232   \n",
       "121837        GACAT       0.008788           0.00709        0.00232   \n",
       "\n",
       "        N1 Length Max  ...  N3 Length Max  N3 SD Avg  N3 SD Median  N3 SD Min  \\\n",
       "0              0.0339  ...         0.0329   4.386989         3.440      0.773   \n",
       "1              0.0222  ...         0.0262   3.016599         2.660      0.715   \n",
       "2              0.0299  ...         0.0266   2.087146         1.910      0.630   \n",
       "3              0.0370  ...         0.0214   2.236520         2.135      0.884   \n",
       "4              0.0478  ...         0.0485   4.260253         4.160      1.040   \n",
       "...               ...  ...            ...        ...           ...        ...   \n",
       "121833         0.0432  ...         0.0242   4.005616         3.640      1.030   \n",
       "121834         0.0262  ...         0.0149   3.644638         3.440      1.580   \n",
       "121835         0.0315  ...         0.0260   2.181562         1.925      1.040   \n",
       "121836         0.0483  ...         0.0144   2.540877         2.330      1.190   \n",
       "121837         0.0215  ...         0.0306   2.576731         2.295      1.330   \n",
       "\n",
       "        N3 SD Max  N3 Mean Avg  N3 Mean Median  N3 Mean Min  N3 Mean Max  \\\n",
       "0           15.50    80.570270           80.50         73.1         88.3   \n",
       "1           14.10    94.290698           94.10         88.6        103.0   \n",
       "2            6.85    89.364324           89.20         84.4         96.2   \n",
       "3            6.49    89.154000           89.90         81.4         95.7   \n",
       "4            8.81    85.178788           85.40         77.6         90.5   \n",
       "...           ...          ...             ...          ...          ...   \n",
       "121833      12.50    82.004110           82.10         74.0         92.8   \n",
       "121834       6.50    80.497101           80.60         75.0         90.6   \n",
       "121835       7.19    84.190625           84.50         78.2         88.1   \n",
       "121836       6.50    82.289474           82.00         77.6         87.4   \n",
       "121837       5.67    78.536538           77.45         74.4         87.6   \n",
       "\n",
       "        Label  \n",
       "0           0  \n",
       "1           0  \n",
       "2           0  \n",
       "3           0  \n",
       "4           0  \n",
       "...       ...  \n",
       "121833      1  \n",
       "121834      0  \n",
       "121835      1  \n",
       "121836      0  \n",
       "121837      0  \n",
       "\n",
       "[121838 rows x 43 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df = pd.concat([labels_df['gene_id'], data_df, labels_df['label']], axis = 1)\n",
    "final_df.rename(columns={'gene_id': 'Gene ID', 'label': 'Label'}, inplace=True)\n",
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = os.path.join(cwd, 'parsed_data_final.csv')\n",
    "final_df.to_csv(output_path)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
